scripts/build-tools.sh

```diff
@@ -5,69 +5,112 @@
 # fail immediately on any errors
 set -e
 
-
 print_usage()
 {
         cat <<EOFUSAGE
-usage: $0 [-t|-f]
+usage: $0 [-c|-f|-h|-l|-p|-t|-T]
+       [-c] Build ctl
+       [-f] Build fuzzer
+       [-h] Display help
+       [-l] Build logger
+       [-p] Build probes
        [-t] Build test topologies
-       [-f] Build fuzzer"
+       [-T] Build topologies
 EOFUSAGE
 }
 
-build_tools()
+# generate Makefiles
+prepare_build()
 {
-        cd "$SOF_REPO/tools"
-        rm -rf build_tools
-        mkdir build_tools
-        cd build_tools
+        if ! [ -d "$BUILD_TOOLS_DIR" ]
+        then
+                mkdir "$BUILD_TOOLS_DIR"
+        fi
```

Comment:
tl;dr: keep the `rm -rf build_tools`.

This script is rebuilding from scratch and is documented as such
https://thesofproject.github.io/latest/getting_started/build-guide/build-from-scratch.html#step-4-build-topology-and-tools

This change is converting it to incremental builds which is very big change.

In an ideal world, building from scratch is never required. In the real world, incremental builds are frequently broken.

There's also the case of people working on the build system itself. `cmake` tries hard to detect CMakeLists.txt changes and to reconfigure itself but there is only so much it can do.

Anyone working on the code itself (as opposed to the build system) and needing the fastest build time can and should invoke [c]make (or ninja) directly, they don't need and shouldn't use this script. Removing a layer of indirection also allows finer-grained control on the build, think `cmake -DFOO=bar`, `make VERBOSE=1`, etc.

I understand `cmake` and `make` offer a slightly less user-friendly interface but I don't think  that's a huge price to pay if you want incremental builds. One possible way to make  `cmake` and `make` easier could be this, up to you:

```shell
build_tool()
{
    tool=$1
    ( set -x
       make -C ....
    )
}
```
This cheap logging makes it possible to just copy/paste the command for re-use.
It doesn't hurt to see the exact command in CI either.

Reply:
Yes, docs seem to already cover and recommend the approach you stated. However, if we remove build_tools everytime, having individual switch to build individual tools might not make sense, because if I run `./scripts/build-tools -f`, I get fuzzer, if I do `./scripts/build-tools -p` I will get probes, but the fuzzer will be removed. At any given moment it will only contain tools that were last built by the script.

Reply:
TODO

Reply:
I spent a bit more time looking at this. `rm -rf "$BUILD_TOOLS_DIR/$tool"` is not guaranteed to work because the top-level CMakeLists.txt file does not promise that every $tool will be built in a subdirectory name $tool.

> because if I run ./scripts/build-tools -f, I get fuzzer, if I do ./scripts/build-tools -p I will get probes, but the fuzzer will be removed. At any given moment it will only contain tools that were last built by the script.

It depends, one tool/target could depend on another, so you could have more than one.

This made me realize something else: so you apparently want to rebuild one tool without touching the others yet `cmake ..` is re-run at the top-level for ALL the tools!
- `cmake` + `make` for the tools requested
- `cmake` but no make for the tools not requested

That's inconsistent and not every intuitive.

Trying this locally I also found that building everything takes less than 3 seconds on my system.  I also wonder how you tested this because I found the make targets to be called `sof-logger`, `sof-ctl` and `sof-probes`. The  targets `logger` and `ctl` exist but they don't build anything.

I'm wondering what problem(s) this really solves.

Reply:
> This made me realize something else: so you apparently want to rebuild one tool without touching the others yet `cmake ..` is re-run at the top-level for ALL the tools!
> 
> * `cmake` + `make` for the tools requested
> * `cmake` but no make for the tools not requested

Initially I tried to cmake only the tool asked to build. But when I tried 
```
rm -rf tools/build_tools/
mkdir tools/build_tools/logger
cd tools/build_tools/logger
cmake ../../logger
```
the makefiles were written to `tools/logger`, weird. So I tried `cmake -S tools/logger -B tools/build_tools/logger` which worked and was used initially in my PR. However, that is not supported by the cmake in our CI and the README also suggests CMake > 3.10 hence, only way I found was to cmake the whole tools directory and make it work.

> Trying this locally I also found that building everything takes less than 3 seconds on my system. I also wonder how you tested this because I found the make targets to be called `sof-logger`, `sof-ctl` and `sof-probes`. The targets `logger` and `ctl` exist but they don't build anything.

I didn't test make command inside in the `tools/build_tools` directory. I just tested the script, checked the executables in the build directories and even tested make commands in the individual build directories. Does the targets have to be named `logger`, `ctl`, etc? 

> I'm wondering what problem(s) this really solves.

Only specified tools are built, not other tools.

Reply:
@marc-hb we are trying to make the script more flexible so only you compile the things you need. For example, for me it is always a problem when I try to compile the topologies because I always need to export the LD_LIBRARY_PATH to a newer version of alsa-lib so that topology can compile fine.

So, for example I only want to build the fuzzer and not to deal with building the topologies.